File: 0045-jump-game-ii/0045-jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
       int n = nums.length;
        int jumps = 0;         // Number of jumps taken
        int currentEnd = 0;    // End of the current jump range
        int farthest = 0;      // Farthest index reachable

        for (int i = 0; i < n - 1; i++) {
            // Update the farthest index reachable
            farthest = Math.max(farthest, i + nums[i]);

            // If we reach the end of the current jump range, increment jumps
            if (i == currentEnd) {
                jumps++;
                currentEnd = farthest; // Update the end of the next jump range
            }
        }

        return jumps;
    }
}

File: 0045-jump-game-ii/0045-jump-game-ii.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int n=nums.size();
        int curFar=0,curEnd=0;
        int ans=0;
        for(int i=0;i<n-1;i++){
            curFar=max(curFar,i+nums[i]);
            if(i==curEnd){
                curEnd=curFar;
                ans++;
            }
        }
        return ans;
    }
};